pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -81,126 +90,94 @@ func NewDRAPluginClient(pluginName string) (drapb.NodeClient, error) {
 	}
 
+	conn, err := existingPlugin.getOrCreateGRPCConn()
+	if err != nil {
+		return nil, err
+	}
+
 	return &draPluginClient{
-		pluginName:        pluginName,
-		addr:              draAddr(existingPlugin.endpoint),
-		nodeClientCreator: newNodeClient,
+		pluginName:      pluginName,
+		plugin:          existingPlugin,
+		versionedClient: detectVersion(conn),
 	}, nil
 }
 
 func (r *draPluginClient) NodePrepareResources(
 	ctx context.Context,
-	req *drapb.NodePrepareResourcesRequest,
+	req *drapbv1alpha3.NodePrepareResourcesRequest,
 	opts ...grpc.CallOption,
-) (resp *drapb.NodePrepareResourcesResponse, err error) {
+) (response *drapbv1alpha3.NodePrepareResourcesResponse, err error) {
 	logger := klog.FromContext(ctx)
 	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
-	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
+	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", response, "err", err)
 
-	if r.nodeClientCreator == nil {
-		return nil, errors.New("failed to call NodePrepareResources. nodeClientCreator is nil")
-	}
-
-	nodeClient, nodeClientOld, closer, err := r.nodeClientCreator(r.addr)
+	conn, err := r.plugin.getOrCreateGRPCConn()
 	if err != nil {
 		return nil, err
 	}
-	defer closer.Close()
 
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
-	resp, err = nodeClient.NodePrepareResources(ctx, req)
-	if err != nil {
-		status, _ := grpcstatus.FromError(err)
-		if status.Code() == grpccodes.Unimplemented {
-			// Fall back to the older gRPC API.
-			resp = &drapb.NodePrepareResourcesResponse{
-				Claims: make(map[string]*drapb.NodePrepareResourceResponse),
-			}
-			err = nil
-			for _, claim := range req.Claims {
-				respOld, errOld := nodeClientOld.NodePrepareResource(ctx,
-					&drapbv1alpha2.NodePrepareResourceRequest{
-						Namespace:      claim.Namespace,
-						ClaimUid:       claim.Uid,
-						ClaimName:      claim.Name,
-						ResourceHandle: claim.ResourceHandle,
-					})
-				result := &drapb.NodePrepareResourceResponse{}
-				if errOld != nil {
-					result.Error = errOld.Error()
-				} else {
-					result.CDIDevices = respOld.CdiDevices
-				}
-				resp.Claims[claim.Uid] = result
-			}
+	if r.versionedClient.version == "v1alpha3" {
+		response, err = r.versionedClient.v1alpha3NodePrepareResources(ctx, conn, req)
+		return
+	}
+
+	for _, claim := range req.Claims {
+		res, currentError := r.versionedClient.v1alpha2NodePrepareResource(ctx, conn, &drapbv1alpha2.NodePrepareResourceRequest{
+			Namespace:      claim.Namespace,
+			ClaimUid:       claim.Uid,
+			ClaimName:      claim.Name,
+			ResourceHandle: claim.ResourceHandle,
+		})
+		result := &drapbv1alpha3.NodePrepareResourceResponse{}
+		if err != nil {
+			result.Error = currentError.Error()
+		} else {
+			result.CDIDevices = res.CdiDevices
 		}
```

Comment:
I don't follow the logic here with err vs. currentError. At the point we get here, isn't `err` guaranteed to be `nil`?

Independent of that I am strongly against having named return values (which apparently slipped through my previous review of this code).

Reply:
done in 832264ac0d4